### ISSUE TITLE
`panic: incorrect alignment` in Buffer getMappedRange call

### ISSUE THREAD

**Issue by jmrico01** [hexops/mach#847]
Using [this mach version](https://github.com/hexops/mach/tree/1abb7a52dcc96ee3f21d3feccf5c857cd27b186e) with the suggested Zig version `dev.3380`.

I'm doing the following
```zig
const vertexBuffer = core.device().createBuffer(&.{
    .usage = .{.vertex = true},
    .size = @sizeOf(T) * numVertices,
    .mapped_at_creation = true,
});
const mapped = vertexBuffer.getMappedRange(T, 0, numVertices) orelse unreachable;
// ...
```
where `@alignOf(T) == 16`. The call to `.getMappedRange`, results in a panic with `incorrect alignment`, which probably happens during the `@alignCast` inside that function.

Of course, the API has no knowledge of `T` or any alignment requirements upon creation/mapping of the buffer, so this is not too surprising. This seems strange and risky, though. The only workaround I can think of at the moment is to allocate my own `T` array temporarily, then convert it to a byte slice and copy that to the buffer. I say it's strange because this makes me not trust `.getMappedRange` anymore, and I'm sure gonna miss it because it was providing me the convenience of accessing the mapped bytes directly as a `T` array.

Is there some max alignment I can reliably use this for? Maybe 16 is too large? Or is there a way to request a specific alignment on creation/mapping? Couldn't find any hints of that in the source code.

**Comment by slimsag** [comment]
It's my understanding that the alignment of pointers returned by `getMappedRange` in WebGPU is not standardized yet (i.e. the behavior is Dawn-specific); we may need to ask @kainino0x if there have been any developments here - or what Dawn's behavior is today.

I'm aware of two issues that touched on this topic, but didn't provide very clear answers to me aside from 'the returned pointers are probably 16-256 byte aligned':

* https://github.com/webgpu-native/webgpu-headers/issues/180
* https://github.com/emscripten-core/emscripten/pull/19477

There was also some discussion about it in [wgpu-rs](https://github.com/gfx-rs/wgpu/issues/3508#issuecomment-1438271076) being 8-aligned, but I don't know if Dawn has the same requirement today: https://github.com/webgpu-native/webgpu-headers/issues/180

---

What's the stack trace with the panic, out of curiosity? And an example `T` declaration if you have one?

**Comment by jmrico01** [comment]
The stack trace isn't very helpful actually, it doesn't point to the `@alignCast`, just to my function call. I guess that's because `getMappedRange` is an inline function?
```zig
thread 16508 panic: incorrect alignment
C:\<src-path.zig>:27:91: 0x7ff783c17513 in initFromData (zombies.exe.obj)
            var vertexBufferMapped = vertexBuffer.getMappedRange(VertexMach, 0, numPoints * 2) orelse unreachable;

... some of my code here ...

C:\<project-path>\mach\libs\core\src\platform\native\entry.zig:27:17: 0x7ff783c1db9b in main (zombies.exe.obj)
    try app.init();
C:\<zig-path>\lib\std\start.zig:508:80: 0x7ff783c1e161 in main (zombies.exe.obj)
    return @call(.always_inline, callMainWithArgs, .{ @intCast(usize, c_argc), @ptrCast([*][*:0]u8, c_argv), envp });
                                                                               ^
???:?:?: 0x7ff783d529c4 in ??? (zombies.exe)
???:?:?: 0x7ff783d5271b in ??? (zombies.exe)
???:?:?: 0x7ffda65e7613 in ??? (KERNEL32.DLL)
???:?:?: 0x7ffda83a26f0 in ??? (ntdll.dll)
```

and `T` in this case is `VertexMach`, defined as
```zig
pub const VertexMach = struct {
    posDir: @Vector(4, f32),
    prev: @Vector(3, f32),
    next: @Vector(3, f32),
    thickness: f32,
    color: @Vector(4, f32),
};
```
which is 16-aligned because of my 4-wide `f32` fields. Those actually aren't super necessary, so I might swap them to structs with 4 `f32`s instead.

**Comment by jmrico01** [comment]
TL DR I am dumb and was creating buffers with `.size = 0`. Panic messages led me to overthink this.

**Comment by slimsag** [comment]
OOooh, that's really interesting. So what is your final assessment of the alignment requirements? I assume 16 bytes works now?

We can add a check for the `.size = 0` foot-gun here, I'm sure more people will encounter that.

**Comment by jmrico01** [comment]
Yep, 16 byte alignment works !

**Comment by kainino0x** [comment]
> It's my understanding that the alignment of pointers returned by `getMappedRange` in WebGPU is not standardized yet (i.e. the behavior is Dawn-specific); we may need to ask @kainino0x if there have been any developments here - or what Dawn's behavior is today.

There aren't further developments yet. Personally, I don't actually know what guarantees we can get out of the underlying native APIs.

I'll also note that getMappedRange won't guarantee the alignment on its own. For example if you getMappedRange with an offset of 8, you shouldn't expect something that's 16-byte aligned. Most likely what's guaranteed to be aligned is (the pointer you get, minus your getMappedRange offset, minus your mapAsync offset) or something along those lines.
